File: engine/src/main/java/de/hanno/hpengine/engine/scene/AABB.java
package de.hanno.hpengine.engine.scene;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

import de.hanno.hpengine.engine.camera.Camera;

import de.hanno.hpengine.engine.entity.Entity;
import de.hanno.hpengine.log.ConsoleLogger;
import org.joml.Vector3f;
import org.joml.Vector4f;

public class AABB implements Serializable {
	private static Logger LOGGER = ConsoleLogger.getLogger();

	private Vector3f absoluteMaximum = new Vector3f(java.lang.Float.MAX_VALUE, java.lang.Float.MAX_VALUE, java.lang.Float.MAX_VALUE);
	private Vector3f absoluteMinimum = new Vector3f(-java.lang.Float.MAX_VALUE, -java.lang.Float.MAX_VALUE, -java.lang.Float.MAX_VALUE);

	// this is the point -x, y, z if you look in -z with opengl coords
	private Vector3f topRightForeCorner;
	private Vector3f bottomLeftBackCorner;
	private Vector3f center;
	public float sizeX;
	public float sizeY;
	public float sizeZ;

    public Vector3f getMin() { return bottomLeftBackCorner; }
    public Vector3f getMax() { return topRightForeCorner; }

	public AABB(Vector3f center, float size) {
		this(center, size, size, size);
	}
	
	public AABB(Vector3f center, float sizeX, float sizeY, float sizeZ) {
		this.sizeX = sizeX;
		this.sizeY = sizeY;
		this.sizeZ = sizeZ;
		this.center = center;
		calculateCorners();
	}
	
	public void move(Vector3f amount) {
		center.add(amount);
		calculateCorners();
	}

	public void setSize(float size) {
		setSize(size, size, size);
	}
	public void setSize(float sizeX, float sizeY, float sizeZ) {
		this.sizeX = sizeX;
		this.sizeY = sizeY;
		this.sizeZ = sizeZ;
		calculateCorners();
	}
	private void calculateCorners() {
		float halfX = sizeX/2;
		float halfY = sizeY/2;
		float halfZ = sizeZ/2;
		this.bottomLeftBackCorner = new Vector3f(center.x - halfX, center.y - halfY, center.z - halfZ);
		this.topRightForeCorner = new Vector3f(center.x + halfX, center.y + halfY, center.z + halfZ);
	}
	
	public List<Vector3f> getPoints() {
		List<Vector3f> result = new ArrayList<>();
		
		result.add(bottomLeftBackCorner);
		result.add(new Vector3f(bottomLeftBackCorner.x + sizeX, bottomLeftBackCorner.y, bottomLeftBackCorner.z));
		result.add(new Vector3f(bottomLeftBackCorner.x + sizeX, bottomLeftBackCorner.y, bottomLeftBackCorner.z + sizeZ));
		result.add(new Vector3f(bottomLeftBackCorner.x, bottomLeftBackCorner.y, bottomLeftBackCorner.z + sizeZ));
		
		result.add(topRightForeCorner);
		result.add(new Vector3f(topRightForeCorner.x - sizeX, topRightForeCorner.y, topRightForeCorner.z));
		result.add(new Vector3f(topRightForeCorner.x - sizeX, topRightForeCorner.y, topRightForeCorner.z - sizeZ));
		result.add(new Vector3f(topRightForeCorner.x, topRightForeCorner.y, topRightForeCorner.z - sizeZ));

		return result;
	}
	
	public float[] getPointsAsArray() {
		List<Vector3f> points = getPoints();
		List<Vector3f> pointsForLineDrawing = new ArrayList<>();
		pointsForLineDrawing.add(points.get(0));
		pointsForLineDrawing.add(points.get(1));
		pointsForLineDrawing.add(points.get(1));
		pointsForLineDrawing.add(points.get(2));
		pointsForLineDrawing.add(points.get(2));
		pointsForLineDrawing.add(points.get(3));
		pointsForLineDrawing.add(points.get(3));
		pointsForLineDrawing.add(points.get(0));
		
		pointsForLineDrawing.add(points.get(4));
		pointsForLineDrawing.add(points.get(5));
		pointsForLineDrawing.add(points.get(5));
		pointsForLineDrawing.add(points.get(6));
		pointsForLineDrawing.add(points.get(6));
		pointsForLineDrawing.add(points.get(7));
		pointsForLineDrawing.add(points.get(7));
		pointsForLineDrawing.add(points.get(4));

		pointsForLineDrawing.add(points.get(0));
		pointsForLineDrawing.add(points.get(6));
		pointsForLineDrawing.add(points.get(1));
		pointsForLineDrawing.add(points.get(7));
		pointsForLineDrawing.add(points.get(2));
		pointsForLineDrawing.add(points.get(4));
		pointsForLineDrawing.add(points.get(3));
		pointsForLineDrawing.add(points.get(5));

		float[] dest = new float[3* pointsForLineDrawing.size()];
		for (int i = 0; i < pointsForLineDrawing.size(); i++) {
			dest[3*i] = pointsForLineDrawing.get(i).x;
			dest[3*i+1] = pointsForLineDrawing.get(i).y;
			dest[3*i+2] = pointsForLineDrawing.get(i).z;
		}
		return dest;
	}

	public boolean contains(Vector4f point) {
		// max x of box = toprightforecorner.x
		// min x of box = bottomLeftBackCorner.x etc
		
		if (point.x >= bottomLeftBackCorner.x &&
			point.x <= topRightForeCorner.x &&
			
			point.y >= bottomLeftBackCorner.y &&
			point.y <= topRightForeCorner.y &&
			
			point.z >= bottomLeftBackCorner.z &&
			point.z <= topRightForeCorner.z) {
			return true;
		}
		return false;
	}

	public boolean contains(Vector3f position) {
		Vector4f point = new Vector4f(position.x, position.y, position.z, 0);
		return contains(point);
	}

	@Override
	public String toString() {
		return String.format("Box (%f %f %f) @ (%.2f, %.2f, %.2f)", sizeX, sizeY, sizeZ, center.x, center.y, center.z);
	}
	
	@Override
	public boolean equals(Object other) {
		if (!(other instanceof AABB)) {
			return false;
		}
		AABB otherBox = (AABB) other;
		return (otherBox.center.equals(center)
				&& otherBox.sizeX == sizeX
				&& otherBox.sizeY == sizeY
				&& otherBox.sizeZ == sizeZ);
	}

	public Vector3f getTopRightForeCorner() {
		return topRightForeCorner;
	}

	public Vector3f getBottomLeftBackCorner() {
		return bottomLeftBackCorner;
	}


	public boolean isInFrustum(Camera camera) {
		Vector3f centerWorld = new Vector3f();
		topRightForeCorner.add(bottomLeftBackCorner, centerWorld);
		centerWorld.mul(0.5f);
		
		//if (de.hanno.hpengine.camera.getFrustum().cubeInFrustum(centerWorld.x, centerWorld.y, centerWorld.z, size/2)) {
		if (camera.getFrustum().sphereInFrustum(centerWorld.x, centerWorld.y, centerWorld.z, Math.max(sizeX, Math.max(sizeY, sizeZ))/2f)) {
			return true;
		}
		return false;
	}
	
	// TODO: Fix
	public boolean containsOrIntersectsSphere(Vector3f position, float radius) {
		boolean result = false;
		if (this.contains(new Vector4f(position.x, position.y, position.z, 0))) {
			result = true;
			return result;
		}
		List<Vector3f> points = getPoints();
		float smallestDistance = smallestDistance(points, position);
		float largestDistance = largestDistance(points, position);
		if (largestDistance <= radius) {
			result = true;
		}
		return result;
	}

	private float smallestDistance(List<Vector3f> points, Vector3f pivot) {
		float length = Float.MAX_VALUE;
		for (Vector3f point : points) {
			float tempLength = new Vector3f(point).sub(pivot).length();
			length = tempLength <= length? tempLength : length;
		}
		
		return length;
	}
	private float largestDistance(List<Vector3f> points, Vector3f pivot) {
		float length = Float.MAX_VALUE;
		for (Vector3f point : points) {
			float tempLength = new Vector3f(point).sub(pivot).length();
			length = tempLength >= length? tempLength : length;
		}
		
		return length;
	}

	public Vector3f getCenter() {
		return center;
	}

	public void setCenter(Vector3f center) {
		this.center = center;
		calculateCorners();
	}

	public void calculateMinMax(List<Entity> entities) {
		if (entities.isEmpty()) {
			getMin().set(-1f, -1f, -1f);
			getMax().set(1f, 1f, 1f);
			return;
		}

		getMin().set(absoluteMaximum);
		getMax().set(absoluteMinimum);

		for (Entity entity : entities) {
			de.hanno.hpengine.engine.transform.AABB minMaxWorld = entity.getMinMaxWorld();
			Vector3f currentMin = minMaxWorld.getMin();
			Vector3f currentMax = minMaxWorld.getMax();
			getMin().x = (currentMin.x < getMin().x) ? currentMin.x : getMin().x;
			getMin().y = (currentMin.y < getMin().y) ? currentMin.y : getMin().y;
			getMin().z = (currentMin.z < getMin().z) ? currentMin.z : getMin().z;

			getMax().x = (currentMax.x > getMax().x) ? currentMax.x : getMax().x;
			getMax().y = (currentMax.y > getMax().y) ? currentMax.y : getMax().y;
			getMax().z = (currentMax.z > getMax().z) ? currentMax.z : getMax().z;
		}
	}
}

File: engine/src/main/java/de/hanno/hpengine/engine/transform/AABB.kt
package de.hanno.hpengine.engine.transform

import de.hanno.hpengine.engine.camera.Camera
import de.hanno.hpengine.engine.component.ModelComponent
import de.hanno.hpengine.engine.entity.Entity
import org.joml.Vector3f
import org.joml.Vector3fc
import org.joml.Vector4f
import java.util.ArrayList

data class AABB(var min: Vector3f = Vector3f(), var max: Vector3f = Vector3f()) {
    val extents: Vector3f
            get() = Vector3f(max).sub(min)
    val halfExtents: Vector3f
            get() = Vector3f(extents).mul(0.5f)

    constructor(center: Vector3f, halfExtents: Float):
            this(Vector3f(center).sub(Vector3f(halfExtents)), Vector3f(center).add(Vector3f(halfExtents)))

    private val  corners = Array(8) { Vector3f() }

    fun transform(transform: Transform<*>, minMaxWorldProperty: AABB) {
        transform.transformPosition(min, corners[0])
        transform.transformPosition(corners[1].set(min.x, min.y, max.z), corners[1])
        transform.transformPosition(corners[2].set(max.x, min.y, min.z), corners[2])
        transform.transformPosition(corners[3].set(max.x, min.y, max.z), corners[3])

        transform.transformPosition(max, corners[4])
        transform.transformPosition(corners[5].set(min.x, max.y, max.z), corners[5])
        transform.transformPosition(corners[6].set(max.x, max.y, min.z), corners[6])
        transform.transformPosition(corners[7].set(min.x, max.y, min.z), corners[7])

        minMaxWorldProperty.resetToAbsoluteMinMax()
        corners.forEach { corner ->
            minMaxWorldProperty.min.min(corner)
            minMaxWorldProperty.max.max(corner)
        }
    }

    private fun resetToAbsoluteMinMax() {
        min.set(Spatial.MIN)
        max.set(Spatial.MAX)
    }

    private fun clear() {
        min.set(0f,0f,0f)
        max.set(0f,0f,0f)
    }

    fun contains(position: Vector3f): Boolean = contains(Vector4f(position.x, position.y, position.z, 1.0f))
    fun contains(position: Vector4f): Boolean {
        return min.x < position.x && max.x > position.x &&
                min.y < position.y && max.y > position.y &&
                min.z < position.z && max.z > position.z
    }


    fun getPoints(): List<Vector3f> {
        val result: MutableList<Vector3f> = ArrayList()
        result.add(min)
        result.add(Vector3f(min.x + extents.x, min.y, min.z))
        result.add(Vector3f(min.x + extents.x, min.y, min.z + extents.z))
        result.add(Vector3f(min.x, min.y, min.z + extents.z))
        result.add(max)
        result.add(Vector3f(max.x - extents.x, max.y, max.z))
        result.add(Vector3f(max.x - extents.x, max.y, max.z - extents.z))
        result.add(Vector3f(max.x, max.y, max.z - extents.z))
        return result
    }
    fun getPointsAsArray(): FloatArray? {
        val points: List<Vector3f> = getPoints()
        val pointsForLineDrawing: MutableList<Vector3f> = ArrayList()
        pointsForLineDrawing.add(points[0])
        pointsForLineDrawing.add(points[1])
        pointsForLineDrawing.add(points[1])
        pointsForLineDrawing.add(points[2])
        pointsForLineDrawing.add(points[2])
        pointsForLineDrawing.add(points[3])
        pointsForLineDrawing.add(points[3])
        pointsForLineDrawing.add(points[0])
        pointsForLineDrawing.add(points[4])
        pointsForLineDrawing.add(points[5])
        pointsForLineDrawing.add(points[5])
        pointsForLineDrawing.add(points[6])
        pointsForLineDrawing.add(points[6])
        pointsForLineDrawing.add(points[7])
        pointsForLineDrawing.add(points[7])
        pointsForLineDrawing.add(points[4])
        pointsForLineDrawing.add(points[0])
        pointsForLineDrawing.add(points[6])
        pointsForLineDrawing.add(points[1])
        pointsForLineDrawing.add(points[7])
        pointsForLineDrawing.add(points[2])
        pointsForLineDrawing.add(points[4])
        pointsForLineDrawing.add(points[3])
        pointsForLineDrawing.add(points[5])
        val dest = FloatArray(3 * pointsForLineDrawing.size)
        for (i in pointsForLineDrawing.indices) {
            dest[3 * i] = pointsForLineDrawing[i].x
            dest[3 * i + 1] = pointsForLineDrawing[i].y
            dest[3 * i + 2] = pointsForLineDrawing[i].z
        }
        return dest
    }

    fun move(amount: Vector3f) {
        min.add(amount)
        max.add(amount)
    }
}


fun AABB.isInFrustum(camera: Camera): Boolean {
    val centerWorld = Vector3f()
    min.add(halfExtents, centerWorld)
    return camera.frustum.sphereInFrustum(centerWorld.x, centerWorld.y, centerWorld.z, halfExtents.x.coerceAtLeast(halfExtents.y.coerceAtLeast(halfExtents.z)) / 2f)
}

// TODO: Fix
fun AABB.containsOrIntersectsSphere(position: Vector3f, radius: Float): Boolean {
    var result = false
    if (contains(Vector4f(position.x, position.y, position.z, 0f))) {
        result = true
        return result
    }
    val points: List<Vector3f> = getPoints()
    val smallestDistance: Float = smallestDistance(points, position)
    val largestDistance: Float = largestDistance(points, position)
    if (largestDistance <= radius) {
        result = true
    }
    return result
}

private fun smallestDistance(points: List<Vector3f>, pivot: Vector3f): Float {
    var length = Float.MAX_VALUE
    for (point in points) {
        val tempLength = Vector3f(point).sub(pivot).length()
        length = if (tempLength <= length) tempLength else length
    }
    return length
}

private fun largestDistance(points: List<Vector3f>, pivot: Vector3f): Float {
    var length = Float.MAX_VALUE
    for (point in points) {
        val tempLength = Vector3f(point).sub(pivot).length()
        length = if (tempLength >= length) tempLength else length
    }
    return length
}

private val absoluteMaximum = Vector3f(Float.MAX_VALUE, Float.MAX_VALUE, Float.MAX_VALUE)
private val absoluteMinimum = Vector3f(-Float.MAX_VALUE, -Float.MAX_VALUE, -Float.MAX_VALUE)

fun AABB.calculateMinMax(entities: List<Entity>) {
    if (entities.isEmpty()) {
        min.set(-1f, -1f, -1f)
        max.set(1f, 1f, 1f)
        return
    }
    min.set(absoluteMaximum)
    max.set(absoluteMinimum)
    for (entity in entities) {
        if (!entity.hasComponent(ModelComponent::class.java)) {
            continue
        }
        val minMaxWorld = entity.minMaxWorld
        val currentMin = minMaxWorld.min
        val currentMax = minMaxWorld.max
        min.x = currentMin.x.coerceAtMost(min.x)
        min.y = currentMin.y.coerceAtMost(min.y)
        min.z = currentMin.z.coerceAtMost(min.z)
        max.x = currentMax.x.coerceAtLeast(max.x)
        max.y = currentMax.y.coerceAtLeast(max.y)
        max.z = currentMax.z.coerceAtLeast(max.z)
    }
}